### ISSUE TITLE
The {RD,WR}{FS,GS}BASE instruction is not currently enabled. Please reload Graphene SGX kernel module.

### ISSUE THREAD

**Issue by markueni** [oscarlab/graphene#1832]
Hei,

I had ported one of my applications into Graphene-SGX, and now when I came back to run it as:

`SGX=1 ./pal_loader $(my_executable)`

I got the following error

`The {RD,WR}{FS,GS}BASE instruction is not currently enabled. Please reload Graphene SGX kernel module.`

However I have loaded Graphene SGX kernel module as it can be seen below
```
gsgx                   16384  0
intel_sgx              53248  0
```
May I know what is going wrong?

Is there a new version of Graphene which I need to clone and built it again, or is something else going wrong?

**Comment by mkow** [comment]
Closing. Please comment if the instructions I pasted didn't help.

**Comment by ThibsG** [comment]
I am hitting the same issue running a fresh installed Ubuntu 18.04.
As far as you know, does this Linux update (preventing the enabling of `FSGSBASE`) is applied to old versions as-well (like kernel 5.4 for Ubuntu 18.04) or only to the latest version?
Am I right to think that now the `2b. Install the Graphene FSGSBASE driver (not for production)` is not working anymore and we are forced to patch the kernel or use one >= v5.9?

**Comment by dimakuv** [comment]
https://graphene.readthedocs.io/en/latest/building.html#a-install-the-linux-kernel-patched-with-fsgsbase -- there are instructions how to patch the Linux 5.4 kernel with the FSGSBASE patch.

Also (and **insecure, at your own risk**), you can try to use the **only-for-testing** kernel module that enables FSGSBASE **insecurely**. It is still available here: https://github.com/oscarlab/graphene-sgx-driver/ (see the end of the README). But this is **deprecated** because it is unsafe.

**Comment by ThibsG** [comment]
Yes ok, so the Graphene hacky-driver should still work..
Hmm not sure why I'm getting this message though.
Thanks, I will continue to find a way to make it work

**Comment by dimakuv** [comment]
@ThibsG If you're getting this code, most probably the Linux kernel contains a "someone set an unsafe FSGSBASE feature, I will unset it now" (there was a patch that was backported to Linux 5.4 among others, and I guess Ubuntu 18.04 has this patch by default now). You can check what happens in the kernel via `dmesg`. Check for FSGSBASE-related messages in that log.

**Comment by ThibsG** [comment]
> If you're getting this code, most probably the Linux kernel contains a "someone set an unsafe FSGSBASE feature, I will unset it now" (there was a patch that was backported to Linux 5.4 among others, and I guess Ubuntu 18.04 has this patch by default now).

Yes that's exactly I'm afraid of 😅 

> You can check what happens in the kernel via dmesg. Check for FSGSBASE-related messages in that log.

Hmm nothing relevant from there:
```
[24336.086162] gsgx: Graphene SGX Driver v0.10
[24336.086162] gsgx: ********************************************************************
[24336.086164] gsgx: *** You have loaded a module that is intended only for research  ***
[24336.086164] gsgx: *** and development, and not for use on any production system.   ***
[24336.086165] gsgx: *** This module is known to allow unprivileged applications to   ***
[24336.086165] gsgx: *** gain full control of the operating system kernel. If this is ***
[24336.086166] gsgx: *** a production system you should shut down immediately.        ***
[24336.086167] gsgx: *** For more information see the included README.rst.            ***
[24336.086167] gsgx: ********************************************************************
[24371.289135] intel_sgx: Intel SGX Driver v2.11.0
[24371.289147] intel_sgx INT0E0C:00: EPC bank 0x50200000-0x55f80000
[24371.290094] intel_sgx: second initialization call skipped
```

Maybe a problem with the SGX Driver version?

**Comment by ThibsG** [comment]
Finally I chose to upgrade the kernel to 5.9. I tried to patch my 5.4 kernel following Graphene doc instructions, but it didn't work when compiling patched Linux source code (unknown compile error about certs, googling didn't help), so I used a graphical tool called [Mainline](https://ubuntuhandbook.org/index.php/2020/08/mainline-install-latest-kernel-ubuntu-linux-mint/) and I selected the 5.9.0 kernel. Didn't test the whole workflow yet but enclave is launching fine after reboot.
